src/js/utils/JSONEditorUtil.js

```diff
@@ -0,0 +1,224 @@
+        );
+        return memo;
+      }
+
+      newKeys.splice(i, 1);
+
+      return memo.concat(
+        JSONEditorUtil.deepObjectDiff(
+          oldObj[key], newObj[key], path.concat([key])
+        )
+      );
+    }, []);
+
+    // Process new keys
+    diff = newKeys.reduce(function (memo, key) {
+      memo.push(
+        {path: path.concat([key]), value: newObj[key], previous: undefined}
+      );
+      return memo;
+    }, diff);
+
+    // If something changed, include the current object as changed
+    if (diff.length) {
+      diff.push({path, value: newObj, previous: oldObj});
+    }
+
+    return diff;
+  },
+
+  /**
+   * Find the location of a change between two strings
+   *
+   * @param {string} oldString - The old string
+   * @param {string} newString - The new string
+   * @returns {number} Returns the offset in the string with the first difference
+   */
+  diffLocation(oldString, newString) {
+    for (var i=0, l=Math.min(oldString.length, newString.length); i<l; ++i) {
+      if (oldString[i] !== newString[i]) {
+        return i;
+      }
+    }
+    return -1;
+  },
+
+  /**
+   * Arrange object keys (or array elements) of the given new value in order to
+   * match the ones in the old value.
+   *
+   * @param {any} oldVal - The previous value to match against
+   * @param {any} newVal - The new value to adapt
+   * @returns {any} Returns the newVal with it's element properly arranged
+   */
+  sortObjectKeys(oldVal, newVal) {
+    if (Array.isArray(oldVal) && Array.isArray(newVal)) {
+      let oldValues = oldVal.slice();
+      let newValues = newVal.slice();
+
+      // Keep values that exist in both arrays, in the order they appear in the
+      // `oldValues` array. In the same time, strip these values out of the
+      // `newValues` array in order to keep only the `new` values.
+      let resultVal = oldValues.reduce(function (memo, oldItem) {
+
+        // Check if we have the same or compatible item in the `newValues` array
+        // A "compatible" item is an object that has
+        let i = newValues.findIndex(function (newItem) {
+          if (oldItem === newItem) {
+            return true;
+          }
+
+          if ((typeof oldItem === 'object') && (typeof newItem === 'object') &&
+              doKeysOverlap(oldItem, newItem)) {
+            return true;
+          }
+
+          return deepEqual(oldItem, newItem);
+        });
+
+        // If we don't have that item, don't include it in the list
+        if (i === -1) {
+          return memo;
+        }
+
+        // Push (and recursively sort if appliable) the value
+        memo.push(JSONEditorUtil.sortObjectKeys(oldItem, newValues[i]));
+        newValues.splice(i, 1);
+
+        return memo;
+      }, []);
+
+      // Append new values in the order they appear at the end of the array
+      return resultVal.concat(newValues);
+
+    } else if ((typeof oldVal === 'object') && (typeof newVal === 'object')) {
```

Comment:
#nit These `else`s are not necessary, since you are returning in each of the blocks